lemur/plugins/lemur_gcp/plugin.py

```diff
@@ -0,0 +1,114 @@
+            "helpMessage": "Authentication method to use",
+        },
+        {
+            "name": "vaultMountPoint",
+            "type": "str",
+            "required": False,
+            "helpMessage": "Path to vault secret",
+        },
+        {
+            "name": "serviceAccountTokenPath",
+            "type": "str",
+            "required": False,
+            "helpMessage": "Path to vault secret",
+        }
+    ]
+
+    def upload(self, name, body, private_key, cert_chain, options, **kwargs):
+
+        try:
+            ssl_certificate_body = {
+                "name": self._gcp_name(body),
+                "certificate": body,
+                "description": "",
+                "private_key": private_key,
+            }
+            credentials = self._get_gcp_credentials(options)
+            return self._insert_gcp_certificate(
+                self.get_option("projectID", options),
+                ssl_certificate_body,
+                credentials,
+            )
+
+        except Exception as e:
+            current_app.logger.error(
+                f"Issue with uploading {name} to GCP. Action failed with the following log: {e}",
+                exc_info=True,
+            )
+            raise Exception(f"Issue uploading certificate to GCP: {e}")
+
+    def _insert_gcp_certificate(self, project_id, ssl_certificate_body, credentials):
+        return ssl_certificates.SslCertificatesClient(credentials=credentials).insert(
+            project=project_id, ssl_certificate_resource=ssl_certificate_body
+        )
+
+    def _get_gcp_credentials(self, options):
+        if self.get_option('authenticationMethod', options) == "vault":
+            # make a request to vault for GCP token
+            return self._get_gcp_credentials_from_vault(options)
+        elif self.get_option('authenticationMethod', options) == "serviceAccountToken":
+            if self.get_option('serviceAccountTokenPath', options) is not None:
+                return service_account.Credentials.from_service_account_file(
+                    self.get_option('serviceAccountTokenPath', options)
+                )
+
+        raise Exception("No supported way to authenticate with GCP")
+
+    def _get_gcp_credentials_from_vault(self, options):
+        service_token = hvac.Client(os.environ['VAULT_ADDR']) \
+            .secrets.gcp \
+            .generate_oauth2_access_token(
+            roleset="",
+            mount_point=f"{self.get_option('vaultMountPoint', options)}"
+        )["data"]["token"].rstrip(".")
+        credentials, _ = google.auth.default()  # Fetches default GCP credentials for the current environment
+        credentials.token = service_token  # replace default credentials with oauth2 token fetched from vault
+
+        return credentials
+
+    def _gcp_name(self, body):
+        """
+        We need to change the name of the certificate that we are uploading to comply with GCP naming standards.
```

Comment:
There's apparently two more GCP naming constraints which need to be handled here:

- certificate names must be all lower case (the `authority` in this case can contain upercase characters)
- no longer than 63 characters (names longer than 63 characters can be generated)

Reply:
added the new naming constraints as part of this commit: https://github.com/DataDog/lemur/pull/15/commits/e72d08d8d880ee926fa59d6d6c38c9b2f04c815f
I moved the logic for modifying the cert name to its own method for easier testing.

Reply:
follow up. Some uploads were still failing because trailing characters in the name must be alpha numeric. Fixed this issue with https://github.com/DataDog/lemur/pull/15/commits/f02f4743d318607bafa452cfeb36d01125889904